java-checks/src/main/java/org/sonar/java/checks/RedundantModifierCheck.java

```diff
@@ -47,35 +47,71 @@
-        checkRedundantModifier(((MethodTree) member).modifiers(), Modifier.PRIVATE);
+      switch (member.kind()) {
+        case METHOD:
+          checkMethod((MethodTree) member, classTree);
+          break;
+        case VARIABLE:
+          checkVariable((VariableTree) member, classTree);
+          break;
+        case CONSTRUCTOR:
+          if (tree.is(Kind.ENUM)) {
+            checkRedundantModifier(((MethodTree) member).modifiers(), Modifier.PRIVATE);
+          }
+          break;
+        case INTERFACE:
+          ClassTree nestedClass = (ClassTree) member;
+          checkNestedInterface(nestedClass, classTree);
+          checkNestedClass(nestedClass, classTree);
+          break;
+        case CLASS:
+          checkNestedClass((ClassTree) member, classTree);
+          break;
+        default:
+          // Do nothing for others members
       }
     }
   }
 
+  private void checkMethod(MethodTree methodTree, ClassTree classTree) {
+    ModifiersTree modifiers = methodTree.modifiers();
+    if (isInterfaceOrAnnotation(classTree)) {
+      checkRedundantModifier(modifiers, Modifier.ABSTRACT, Modifier.PUBLIC);
+    } else if (ModifiersUtils.hasModifier(classTree.modifiers(), Modifier.FINAL)) {
+      checkRedundantModifier(modifiers, Modifier.FINAL);
+    }
+  }
+
+  private void checkVariable(VariableTree variableTree, ClassTree classTree) {
+    if (isInterfaceOrAnnotation(classTree)) {
+      ModifiersTree modifiers = variableTree.modifiers();
+      checkRedundantModifier(modifiers, Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL);
+    }
+  }
+
+  private void checkNestedClass(ClassTree nested, ClassTree classTree) {
+    if (isInterfaceOrAnnotation(classTree)) {
+      ModifiersTree modifiers = nested.modifiers();
+      checkRedundantModifier(modifiers, Modifier.PUBLIC, Modifier.STATIC);
+    }
+  }
+
+  private void checkNestedInterface(ClassTree nested, ClassTree classTree) {
+    if (classTree.is(Kind.CLASS, Kind.ENUM)) {
+      checkRedundantModifier(nested.modifiers(), Modifier.STATIC);
+    }
+  }
+
   private static boolean isInterfaceOrAnnotation(Tree tree) {
     return tree.is(Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
   }
 
-  private void checkRedundantModifier(ModifiersTree modifiersTree, Modifier modifier) {
-    ModifierKeywordTree foundModifier = ModifiersUtils.getModifier(modifiersTree, modifier);
-    if (foundModifier != null) {
-      reportIssue(foundModifier, "\"" + modifier.toString().toLowerCase(Locale.US) + "\" is redundant in this context.");
+  private void checkRedundantModifier(ModifiersTree modifiersTree, Modifier... modifiers) {
+    for (Modifier modifier : modifiers) {
+      ModifierKeywordTree foundModifier = ModifiersUtils.getModifier(modifiersTree, modifier);
+      if (foundModifier != null) {
+        reportIssue(foundModifier, "\"" + modifier.toString().toLowerCase(Locale.US) + "\" is redundant in this context.");
```

Comment:
I'll suggest also to have a String representation of each modifier somewhere in utils or helpers. Cause it's not optimal to have `modifier.toString().toLowerCase(Locale.US)`  each time we need it.

Also, that's up to you, personally I prefer `String.format()` over String concat.

Reply:
I agree for the `String.format()`.

>it's not optimal to have modifier.toString().toLowerCase(Locale.US)

I refactored the code to have `foundModifier.keyword().text()`, I feel that it's slightly better (at least, we do not recompute `toLowerCase`...). 
In any way, we will only call this code when an issue is reported, so it should not be too frequent.

Reply:
LGTM